### PR TITLE
[VOTE PENDING BUIP085 vs BUIP088] Double spend relaying

### PR DESCRIPTION
**Pull request by dagurval**
Double spend relaying ported from Bitcoin XT. Mainly:

https://github.com/bitcoinxt/bitcoinxt/pull/293
https://github.com/bitcoinxt/bitcoinxt/pull/327

Wallet integration is stripped out. QA test depended on wallet integration, so that's also not included. The code is however unit tested.

### CODE REVIEW

**Comment by dgenr8** [comment]
Please squash that last commit of mine, it's just making this harder to review.

**Comment by dagurval** [comment]
Squashed https://github.com/bitcoinxt/bitcoinxt/pull/327 as suggested by @dgenr8. Ran clang-format on files that travis complained about.

**Comment by dagurval** [comment]
Looks like Travis timed out (ping @sickpig)

File: src/respend/respendrelayer.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        RelayLimiter() : respendCount(0), lastRespendTime(0) { }
+
+        bool HasLimitExceeded(const CTransaction& doubleSpend)
+        {
+            unsigned int size = ::GetSerializeSize(doubleSpend,
+                                                   SER_NETWORK, PROTOCOL_VERSION);
+
+            std::lock_guard<std::mutex> lock(cs);
+            int64_t limit = GetArg("-limitrespendrelay", 100);
+            if (RateLimitExceeded(respendCount, lastRespendTime, limit, size)) {
```
Comment (ptschip) [inline src/respend/respendrelayer.cpp]: you'll need to add a line to allowed_args.cpp for -limitspendrelay. We don't allow random entries in the bitcoin.conf, has to be an allowed arg.... you can find the same under limitfreerelay in allowed_args.cpp  like this:          .addDebugArg("limitfreerelay=<n>", optionalInt,
            strprintf("Continuously rate-limit free transactions to <n>*1000 bytes per minute (default: %u)",
                         DEFAULT_LIMITFREERELAY))

Comment (dagurval) [reply src/respend/respendrelayer.cpp]: Thanks! Pushed a commit that adds limitrespendrelay to allowed args.

**Comment by ptschip** [comment]
@dagurval In main.cpp for AlreadyHave()  there is:

        return rrc || mempool.exists(inv.hash) || AlreadyHaveOrphan(inv.hash) ||
               pcoinsTip->HaveCoinInCache(COutPoint(inv.hash, 0)) || // Best effort: only try output 0 and 1
               pcoinsTip->HaveCoinInCache(COutPoint(inv.hash, 1));

It's always bothered me that it checks pcointTip for whether we have this tx or not and was wanting to take it out...  Won't this also prevent us from requesting and fowardding potential double spends?

**Comment by dgenr8** [comment]
@ptschip When evaluating a respend just received, this is looking for coins belonging to the respend itself so it shouldn't be a problem.

**Comment by ptschip** [comment]
maybe i'm missing something but AlreadyHave() is checked both when we
receive and INV and when we AskFor() a txn.  So if a doublespend comes
by we will never ask for it and therefore never forward it on???

**Comment by ptschip** [comment]
@dagurval ok thanks for the explanation on slack...no problems here.

**Comment by ptschip** [comment]
@dgenr8 oops, meant that last comment for you...

**Comment by zander** [comment]
Copying some comments from BUIP85 here;

Double spend relaying was required on the old chain, there was no better solution possible.
Using the BCH chain this is no longer true. It is possible to create double-spend proofs very much like I designed them in 2016, on the plain simple BCH chain. (see BUIP 88 for details).

The propagation of both the original and the double-spending transactions has huge downsides.

* it doesn't help spv nodes. SPV nodes will just ignore a transaction that isn't meant for them.
* it is very easy to abuse, for instance by making the second transaction really large.
* making one of the two transactions fail policy rules, for instance making it zero-fee, still fails to forward the transactions. This incidentally is 90% of the usecases.
* On a larger scale the idea is also really destructive because the current "first seen" rule will no longer work. With this code I can now send my second transaction 2 minutes after the first and it will still appear in the mempool of most nodes.
[edit; to clarify, I mean that it will be offered to all nodes to put in their mempool, and to SPV wallets, whereas before it would not propagate at all]

A double-spend proof should be relatively easy to write, you create a new message-type and SPV nodes can choose to accept that one too.

A double-spend proof has the advantage that it is much smaller and it still is cryptographically verifiable.

A double-spend proof will propagate through the network much like a normal transaction would, warning all nodes. Miners and trackers then have the ability to timestamp transactions as well, based on seen-time, this may be useful.

Maybe someone can work on a BUIP88 implementation instead?

**Comment by dagurval** [comment]
I just want to note that this PR adds a framework for handling double spends and BUIP88 can be added on top of it by implementing a new `RespendAction`.

Relevant [discussion on relaying vs proofs here](https://github.com/bitcoincashorg/spec/pull/57).

**Comment by ptschip** [comment]
needs a rebase

**Comment by dagurval** [comment]
I rebased this, but github hasn't updated this PR to reflect it. Do I need to open a new PR?

**Comment by sickpig** [comment]
@dagurval please open a new one.

What's happening here happened already for #458. It could be related to some merge problem we have back when we released 1.3.0.1 or maybe due to the fact BitcoinUnlimited fork has been detached from  Bitcoin Core on May 1st 2018. 

edit: use dagur correct nickname (@dagurval rather Dagur)

**Comment by dagurval** [comment]
Re-submitted as #1109

**Pull request closed without merging**